toucan_connectors/databricks/databricks_connector.py

```diff
@@ -0,0 +1,131 @@
+    query: constr(min_length=1) = Field(
+        ...,
+        description='You can write a query here',
+        widget='sql',
+    )
+
+
+class DataBricksConnectionError(Exception):
+    """ """
+
+
+class DataBricksUnauthorizedError(Exception):
+    """ """
+
+
+class DatabricksConnector(ToucanConnector):
+    data_source_model: DatabricksDataSource
+    host: str = Field(
+        ...,
+        description='The listening address of your databricks cluster',
+        placeholder='my-databricks-cluster.cloudproviderdatabricks.net',
+    )
+    port: int = Field(
+        ..., description='The listening port of your databricks cluster', placeholder=443
+    )
+    http_path: str = Field(
+        ..., description='Databricks compute resources URL', placeholder='sql/protocolv1/o/xxx/yyy'
+    )
+    pwd: SecretStr = Field(
+        None, description='Your personal access token', placeholder='dapixxxxxxxxxxx'
+    )
+    ansi: bool = False
+    on_demand: bool = Field(
+        False, description="If the cluster is auto terminating, wait for it's start before querying"
+    )
+
+    def _build_connection_string(self) -> str:
+        """For constants see: https://docs.databricks.com/integrations/bi/jdbc-odbc-bi.html"""
+        connection_params = {
+            'Driver': '/opt/simba/spark/lib/64/libsparkodbc_sb64.so',  # This path must be correct when installing
+            # ODBC driver
+            'Host': '127.0.0.1' if self.host == 'localhost' else self.host,
+            'Port': self.port,
+            'HTTPPath': self.http_path,
+            'ThriftTransport': 2,
+            'SSL': 1,
+            'AuthMech': 3,
+            'UID': 'token',
+            'PWD': self.pwd.get_secret_value(),
+        }
+        return ';'.join(f'{k}={v}' for k, v in connection_params.items() if v is not None)
+
+    @staticmethod
+    def _get_details(index: int, status: Optional[bool]):
+        checks = ['Host resolved', 'Port opened', 'Connected to Databricks', 'Authenticated']
+        ok_checks = [(c, True) for i, c in enumerate(checks) if i < index]
+        new_check = (checks[index], status)
+        not_validated_checks = [(c, None) for i, c in enumerate(checks) if i > index]
+        return ok_checks + [new_check] + not_validated_checks
+
+    def get_status(self) -> ConnectorStatus:
+        # Check hostname
+        try:
+            self.check_hostname(self.host)
+        except Exception as e:
+            return ConnectorStatus(status=False, details=self._get_details(0, False), error=str(e))
+        # Check port
+        try:
+            self.check_port(self.host, self.port)
+        except Exception as e:
+            return ConnectorStatus(status=False, details=self._get_details(1, False), error=str(e))
+        try:
+            pyodbc.connect(self._build_connection_string())
+        except pyodbc.InterfaceError as e:
+            details = self._get_details(3, False)
+            details[2] = ('Connected to Databricks', True)
```

Comment:
Is this really needed ? Shouldn't it already be set to `True` if `index` is 3 ?

Reply:
yep, forgot to clean that :+1: